Feature/nlist buffer tuner

b-butler opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

Adds a neighbor list buffer tuner, and an optimization API for HOOMD's solver infrastructure. Tuning a neighbor list's buffer size comes with some caveats and traps such as too short of trigger for TPS averaging, GPU autotuning, etc. I have attempted to document some of these tips and warnings in the documentation. This would apply to any solvers not just the gradient descent with momentum I implemented.
<!-- Describe your changes in detail. -->

## Motivation and context

We no longer provide a buffer for neighbor list object but picking a poor size can result in simulation speeds over halved.
<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #???

## How has this been tested?

Test on the functionality have been added, but no validation tests on TPS have been due to time constraints. We could potentially add tuner tests to the new validation test suite with very cautious settings.
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Added: A new tuner for the neighbor list buffer size ``hoomd.md.tune.NeighborListBuffer``.
Added: New solver infrastructure for optimization problems.
- ``Simulation.initial_timestep`` the timestep on which the last call to ``run`` started.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

b-butler committed:
Add optimizers to solver infrastructure.

Also separates solver code from attr_tuner.

b-butler committed:
Add NeighborList.buffer tuner.

b-butler committed:
Document tips for using neighbor list buffer tuner

b-butler committed:
Add testing for NeighborList Tuner

b-butler committed:
Add __eq__ for _LoggerQuantity

Needed for testing Internal Custom Actions

b-butler committed:
Make hoomd.md.tune.NeighborListBuffer pickleable

b-butler committed:
Update and fix documentation for Solvers

b-butler committed:
Add more accurate interval TPS to NeighborList buffer tuning

joaander commented:
Thanks, I will review this soon.

joaander commented on hoomd/md/tune/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2009-2022 The Regents of the University of Michigan.
```
Add this package to the sphinx documentation.

joaander commented:
I've started reviewing, will complete the review soon. I want to try tuning the buffer for a few test systems.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+    def __call__(self):
+        if self._simulation.timestep == self._last_timestep:
+            return self._last_tps
+        walltime = self._simulation.walltime
```
This will lead to different walltimes on different MPI ranks when sampling during the run, eventually leading to different buffer sizes on the ranks.

The code I wrote in `System.cc` broadcasts the walltime only at the end of `run`. I think the solution is to move the broadcast for walltime (tps as well) to the getter method so that it is always identical on all ranks.

joaander commented:
Am I using this wrong, it quite often gets tock at a very unoptimal buffer of ~0.00.  Here are two traces, one starts at buffe=0.1, the other at 0.4 (which is close to optimal).
```
Simulation.timestep  Simulation.tps    Cell.buffer    md.nlist.Cell.num_builds
      133000           1406.52323        0.00000                1000          
      134000           1715.94940        0.06250                733           
      135000           2501.22683        0.12500                393           
      136000           2722.56046        0.18750                287           
      137000           3062.76366        0.25000                218           
      138000           3246.50002        0.31250                180           
      139000           3188.98316        0.37500                151           
      140000           2968.11725        0.43750                130           
      141000           3392.60487        0.50000                115           
      142000           3034.58331        0.56250                102           
      143000           3015.30289        0.62500                 91           
      144000           2917.41865        0.68750                 81           
      145000           3307.07097        0.75000                 74           
      146000           2991.09856        0.81250                 70           
      147000           2606.91210        0.87500                 62           
      148000           3071.19076        0.93750                 59           
      149000           2660.19769        1.00000                 55           

Start NeighborListBuffer tuning
      150000           2172.91534        0.10000                500           
      151000           2151.21375        0.11000                500           
      152000           2498.99940        0.16000                335           
      153000           2172.01772        0.11000                500           
      154000           1485.73604        0.06000                881           
      155000           2155.25146        0.11000                500           
      156000           1455.02954        0.06000                908           
      157000           1293.00805        0.01000                1000          
      158000           1348.53119        0.00001                1000          
      159000           1343.48772        0.05001                1000          
      160000           2168.99959        0.10001                500           
      161000           1350.10947        0.05001                1000          
      162000           1359.15409        0.00001                1000          
      163000           1352.39593        0.05001                1000          
      164000           2171.81467        0.10001                500           
      165000           1353.50998        0.05001                1000          
      166000           1356.53152        0.00001                1000          
      167000           1347.10373        0.05001                1000          
      168000           2160.38303        0.10001                500           
      169000           1346.64612        0.05001                1000
```

```
Simulation.timestep  Simulation.tps    Cell.buffer    md.nlist.Cell.num_builds
      133000           1406.62423        0.00000                1000          
      134000           1761.19035        0.06250                705           
      135000           2484.54929        0.12500                398           
      136000           2772.61247        0.18750                280           
      137000           3068.57764        0.25000                220           
      138000           3225.68146        0.31250                178           
      139000           3259.47581        0.37500                150           
      140000           2983.59471        0.43750                131           
      141000           3392.91597        0.50000                114           
      142000           2799.37792        0.56250                101           
      143000           3319.62065        0.62500                 91           
      144000           2958.79895        0.68750                 84           
      145000           3316.05048        0.75000                 75           
      146000           3046.48404        0.81250                 69           
      147000           2565.97958        0.87500                 64           
      148000           3094.46960        0.93750                 59           
      149000           2704.11464        1.00000                 53           

Start NeighborListBuffer tuning
      150000           3237.17928        0.40000                144           
      151000           3274.71195        0.44000                132           
      152000           3172.27451        0.39000                146           
      153000           3103.46756        0.34000                169           
      154000           3096.00402        0.29000                199           
      155000           2961.61463        0.24000                236           
      156000           3116.38238        0.29000                199           
      157000           2961.92196        0.24000                237           
      158000           2767.69636        0.19000                278           
      159000           2966.07276        0.24000                236           
      160000           2772.45089        0.19000                277           
      161000           2691.38031        0.14000                339           
      162000           2186.07216        0.09000                500           
      163000           2692.94013        0.14000                338           
      164000           2166.88719        0.09000                500           
      165000           1354.36141        0.04000                1000          
      166000           2168.27931        0.09000                500           
      167000           1348.60037        0.04000                1000          
      168000           1295.20915        0.00001                1000          
      169000           1371.35986        0.00001                1000 
```

Will share the script and input file for this test later.

joaander commented:
I think I found the problem. You estimate TPS using `sim.walltime` and `sim.timestep`. `sim.walltime` resets to 0 at the start of every run and I had this in my script to the tps output would be accurate:
```
for i in range(20):
    sim.run(period)
```

When I use ``sim.run(period*20)``, the results are more stable.

b-butler commented:
@joaander we could document that the tuner requires runs to be uninterrupted, if that would aid usage. I don't see a way not to use `sim.walltime` as we don't want to include the time in-between runs to measure TPS.

joaander commented:
> @joaander we could document that the tuner requires runs to be uninterrupted, if that would aid usage. I don't see a way not to use `sim.walltime` as we don't want to include the time in-between runs to measure TPS.

Right, I think the best option is to document the behavior and don't tune on steps right after a new run starts. Checking if `sim.walltime` decreases isn't the best solution as it may be possible for the user to issue consecutive `run` calls but still have the walltime of each call increase slightly. Best to check if the start timestep of the run increases instead. We don't currently have an API for that, but could add one to complement `final_timestep`: https://hoomd-blue.readthedocs.io/en/v3.1.0/package-hoomd.html#hoomd.Simulation.final_timestep

The alternative would be to have `Simulation` provide some kind of window averaged TPS for convenience. However the API for this gets complex quickly - how many timesteps is it averaged over? Or is it averaged over a period of wall time (like in v2)? If we make these configurable, then users will need to match settings between `Simulation` and the tuner for reliable results. Our current system of providing `walltime` and `timestep` gives users the most flexibility to compute TPS in a manner that they choose, and `Simulation.tps` is convenient to monitor the performance of a single call to `run`.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+    def _get_tps(self):
+        if self._simulation.tps == 0:
+            return None
+        return self._simulation.tps
```
```suggestion
```
Unused.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+        self._simulation = None
+        self._tuned = False
+        self._max_tps = 0
+        self._best_buffer_size = None
```
```suggestion
        self._best_buffer_size = nlist.buffer
```



`None` leads to errors when logging `best_buffer_size`:
```
Start gradient descent tuning
  Cell.buffer    md.tune.NeighborListBuffer.max_tps md.tune.NeighborListBuffer.best_buffer_size
Traceback (most recent call last):
  File "/scratch/joaander/test/hoomd-nlist-buffer/run.py", line 62, in <module>
    sim.run(SAMPLE_PERIOD*20)
  File "/scratch/joaander/build/hoomd/hoomd/simulation.py", line 448, in run
    self._cpp_sys.run(steps_int, write_at_start)
  File "/scratch/joaander/build/hoomd/hoomd/write/table.py", line 301, in act
    self._write_row(output_dict)
  File "/scratch/joaander/build/hoomd/hoomd/write/table.py", line 280, in _write_row
    self.delimiter.join(
  File "/scratch/joaander/build/hoomd/hoomd/write/table.py", line 281, in <genexpr>
    (self._fmt(data[k], headers[k]) for k in headers)))
  File "/scratch/joaander/build/hoomd/hoomd/write/table.py", line 97, in __call__
    return self.format_num(value, column_width)
  File "/scratch/joaander/build/hoomd/hoomd/write/table.py", line 112, in format_num
    min_len_repr = int(log10(max(abs(value), 1))) + 1
TypeError: bad operand type for abs(): 'NoneType'
```

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+
+    @hoomd.logging.log
+    def max_tps(self):
+        return self._max_tps
```
```suggestion
        return self._max_tps*1e3
```
Return TPS in units of timesteps per second.

Might be better to store `1e3` in a private module level constant.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+        return self._max_tps
+
+    @hoomd.logging.log
+    def best_buffer_size(self):
```
Document loggables and properties in the public user-facing documentation of `NeibhborListBuffer`. I only found these by reading the code.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+
+    def _maximum_buffer_post(self, value: float):
+        if self._simulation is not None:
+            self._tunable.domain = (1e-5, value)
```
```suggestion
            self._tunable.domain = (0, value)
```
Users may run in a system of units where 1e-5 is very large value for buffer.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+            target=0.0,
+            get_x=self._get_buffer,
+            set_x=self._set_buffer,
+            domain=(1e-5, self.maximum_buffer),
```
```suggestion
            domain=(0, self.maximum_buffer),
```

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+            tol (`float`, optional): The absolute tolerance for convergence of
+                y (defaults to 1e-5).
+            max_delta (`float`, optional): The maximum iteration step allowed
+                (defaults to 0.05).
```
Users may operate in a system of units where their maximum buffer is less than 0.05, or where steps of 0.05 are far too small to tune effectively. We either need to make this parameter have no default so users can set it appropriately, or set it as a fraction of `maximum_buffer` which users must already set.

b-butler commented:
@joaander I think the answer depends on how robust we want to be. Then `new_walltime < last_walltime` check should work most of the time since if the tuner ran at least once the wall times should be close and if more than once the last wall time should be much greater. We could add a property to the simulation which records on what time step a run started and that would be fool proof on the Python side at least. I don't think we need a new TPS measurement personally. We could make walltime relative to the beginning of the first run and on subsequent runs do something like `m_walltime += time - run_start_time`, and then walltime is monotonically increasing.

joaander commented:
> @joaander I think the answer depends on how robust we want to be. Then `new_walltime < last_walltime` check should work most of the time since if the tuner ran at least once the wall times should be close and if more than once the last wall time should be much greater. 

Yeah, I was thinking of my original test case where each `run` was for the the same number of steps as the period of the tuner. Maybe that is a torture case that no other user would think of doing.

It's up to you how robust you want to be to prevent bug fixes in the future.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+    def __init__(self, simulation):
+        self._simulation = simulation
+        self._last_timestep = simulation.timestep
+        self._last_walltime = simulation.walltime
```
```suggestion
        self._last_timestep = None
        self._last_walltime = None
```
When the tuner is added between two calls to `run`, this records the final walltime of the previous `run` which is not a valid start time for the next run.

To fully fix this problem, also make this change to `System.cc`
```
--- a/hoomd/System.cc
+++ b/hoomd/System.cc
@@ -84,6 +84,8 @@ void System::run(uint64_t nsteps, bool write_at_start)
 
     // initialize the last status time
     m_initial_time = m_clk.getTime();
+    m_last_walltime = 0;
+    m_last_TPS = 0;
```

joaander replied:
Something else might need to be changed. I implemented these locally and sometimes the tuner gets stuck near the initial buffer value and doesn't move much from it.

b-butler committed:
style: Miscellaneous style changes.

b-butler committed:
feat: Add BinaryOptimizer to solvers

The motivation for this feature is to allow for an optimization protocol
that tests a variety of values without getting stuck in a local extrema.

The solver adds a bucket optimization scheme where the domain is divided
into buckets and each bucket is tested for is y value. The best bucket
is chosen after all are tested, and if the desired number of rounds is
met the bucket center is the final optimization. If the number of rounds
is greater than the process repeats with bin extent serving as the new
domain.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,213 @@
+        self._tunable = None
+
+    @property
+    def tuned(self):
```
Under what conditions does the buffer tuner consider itself tuned? These need to be documented.

b-butler committed:
doc: : Fix GradientDescent documentation and building

Adds missing argument descriptions for GradientDescent, and fixes
formatting errors in documentation for sphinx builds.

b-butler committed:
feat: classmethod creating NeighborListBuffer with BinaryOptimizer

b-butler committed:
feat: Add initial_timestep to Simulation loggables

Returns the first step of a run, which provides a fairly robust
mechanism to detect a new run in an action/operation.

b-butler committed:
fix: Stop invalid nlist buffer tuning with sim.initial_timestep

joaander reviewed (changes requested):
Thanks! 

For the first round of review, I focused on bug fixes and attempting to get this working reasonably well. I'm attaching my test script and input file so you can try it out. To run it, you need to merge #1322 and this together in an integration branch, or remove the `num_builds` log quantity from the test script.

Even after applying all my suggested fixes, this still isn't behaving well. For now, all tests are on `brett` using the GPU. I will revisit CPU and MPI testing once we get this working well on the GPU.

For wca (set r_cut in the script to 2**(1/6)) and SAMPLE_PERIOD=10000, it does OK. It spend a while searching near the start point before finally finding something near the max. In the end it takes about the same number of steps to find the max as a brute force search.
```
     buffer            tps           num_builds   
    0.00000         3108.97842         10000      
    0.06250         3747.80731          7100      
    0.12500         5239.27270          3995      
    0.18750         6082.02060          2824      
    0.25000         6680.82252          2185      
    0.31250         7019.73374          1787      
    0.37500         7253.08396          1513      
    0.43750         7329.04932          1308      
    0.50000         7452.03021          1147      
    0.56250         7176.89441          1016      
    0.62500         7030.01907          913       
    0.68750         6969.72189          825       
    0.75000         6799.88058          751       
    0.81250         6358.10253          694       
    0.87500         6254.22985          630       
    0.93750         6244.18452          584       
    1.00000         5934.10702          545       

Start gradient descent tuning
     buffer          max_tps      best_buffer_size
    0.10000             0             0.10000     
    0.11000          4.41182          0.10000     
    0.10969          4.41182          0.10000     
    0.05969          4.42762          0.10969     
    0.03340          4.42762          0.10969     
    0.08340          4.42762          0.10969     
    0.13340          4.46070          0.08340     
    0.18340          5.15535          0.13340     
    0.23340          5.56247          0.18340     
    0.28340          5.93772          0.23340     
    0.33340          6.31743          0.28340     
    0.38340          6.56345          0.33340     
    0.43340          6.80971          0.38340     
    0.47559          6.98609          0.43340     
    0.50860          7.09678          0.47559     
    0.54010          7.20801          0.50860     
    0.56120          7.24330          0.54010     
    0.57536          7.26378          0.56120     
    0.53364          7.26378          0.56120     
    0.49367          7.26378          0.56120  
```

When I switch to LJ (r_cut=2.5, SAMPLE_PERIOD=10000), the problem is worse. In this run, it eventually finds the max, but only after spending a long time near the starting point:
```
     buffer            tps           num_builds   
    0.00000         1423.11588         10000      
    0.06250         1780.86095          7021      
    0.12500         2503.98549          4009      
    0.18750         2778.30030          2811      
    0.25000         3093.42266          2184      
    0.31250         3257.87450          1780      
    0.37500         3098.39376          1506      
    0.43750         3003.34306          1309      
    0.50000         3048.66429          1154      
    0.56250         2789.67616          1017      
    0.62500         2855.46531          908       
    0.68750         2799.21569          826       
    0.75000         2832.10054          750       
    0.81250         2617.78940          688       
    0.87500         2593.82815          635       
    0.93750         2617.38716          580       
    1.00000         2556.12665          543       

Start gradient descent tuning
     buffer          max_tps      best_buffer_size
    0.10000             0             0.10000     
    0.11000          1.80779          0.10000     
    0.11061          1.80900          0.11000     
    0.10231          1.80900          0.11000     
    0.09259          1.81490          0.10231     
    0.08321          1.82215          0.09259     
    0.07528          1.82975          0.08321     
    0.06792          1.83517          0.07528     
    0.10449          1.83517          0.07528     
    0.13448          1.83517          0.07528     
    0.18448          2.23807          0.13448     
    0.23184          2.23807          0.13448     
    0.27580          2.42628          0.23184     
    0.30932          2.60031          0.27580     
    0.34587          2.71010          0.30932     
    0.34899          2.71010          0.30932     
    0.31109          2.71010          0.30932     
    0.26735          2.71270          0.31109     
    0.25803          2.71270          0.31109     
    0.26359          2.71270          0.31109  
```
Also, I don't understand why the max TPS from the gradient tuning is 2.712*1000, but my quick test shows 3257 at buffer ~= 0.3. Where did the other 500 TPS go? In the WCA example `max_tps` more closely matched the other measurement.

Scripts:
[hoomd-nlist-buffer.zip](https://github.com/glotzerlab/hoomd-blue/files/8671043/hoomd-nlist-buffer.zip)

b-butler committed:
feat: Add last_tps loggable for NeighborListBuffer tuner

b-butler committed:
refactor!: Set GradientDescent.maximize default to True

b-butler committed:
test: Add solver tests for Optimizers.

Also refactors the test infrastructure to make it more modular and
extensible.

b-butler committed:
refactor: Misc. fixes to NeighborListBuffer tuner

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
refactor: Broadcast walltime/tps each timestep

b-butler committed:
doc: Add NeighborListBuffer tuner to sphinx docs.

Also fixes any errors associated with the building.

b-butler committed:
refactor: Set initial NeighborListBuffer.best_buffer to nlist buffer

b-butler committed:
refactor!: Change arg default NeighborListBuffer.with_gradient_descent

max_delta is now None by default.

b-butler committed:
refactor: Improve performance of _IntervalTPS

b-butler committed:
refactor: Resest walltime and TPS at start of run

b-butler committed:
feat: method/loggable wrapping to internal custom operations

Improves/enables sphinx documentation for wrapped members.

b-butler committed:
idoc: Document metaclass _AbstractLoggableWithPassthrough

b-butler committed:
doc: Suggest usage of NeighborListBuffer.with_binary

b-butler commented:
Hmm, pre-commit passes locally for me.

b-butler commented:
pre-commit.ci autofix

joaander commented:
> pre-commit.ci autofix

`pre-commit` says: 
> a conflict occurred when applying fixes to the PR. try pulling the upstream branch

Merge `trunk-minor` into this branch and then you should get the same behavior on your system and in the pull request.

joaander commented on hoomd/System.cc:
```diff
@@ -169,22 +171,21 @@ void System::run(uint64_t nsteps, bool write_at_start)
+            bcast(m_last_TPS, 0, m_exec_conf->getMPICommunicator());
+            bcast(m_last_walltime, 0, m_exec_conf->getMPICommunicator());
             }
+#endif
```
Call the broadcast in the `get*` methods. Calling broadcast every step is unnecessary any may slow the simulation.

joaander commented on hoomd/System.cc:
```diff
@@ -169,22 +171,21 @@ void System::run(uint64_t nsteps, bool write_at_start)
-    if (m_sysdef->isDomainDecomposed())
+    // propagate Python exceptions related to signals
+    if (PyErr_CheckSignals() != 0)
         {
-        bcast(m_last_TPS, 0, m_exec_conf->getMPICommunicator());
-        bcast(m_last_walltime, 0, m_exec_conf->getMPICommunicator());
+        throw pybind11::error_already_set();
         }
```
Move this inside the timestep loop per the fix in #1324. Please don't reintroduce #1323.

b-butler replied:
Didn't plan on it, seems a merge went wrong.

joaander commented on hoomd/md/pytest/test_nlist_tuner.py:
```diff
@@ -0,0 +1,122 @@
+            "tol": 1e-3,
+            "max_delta": 0.4
+        }
+        tuner = md.tune.NeighborListBuffer.with_gradient_descent(
```
Add tests for `with_binary`.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,281 @@
+        """bool: Whether the neighbor list buffer is considered tuned.
+
+        Like other tuners, the tuner is considered tune when the specified
+        solver returns ``True`` when solving. See `hoomd.tune` for more
```
I don't understand this. I don't see any references to return values for solves in the `hoomd.tune `documentation`.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,281 @@
+            nlist (hoomd.md.nlist.NeighborList): Neighbor list buffer to
+                maximize TPS.
+            maximum_buffer (float): The largest buffer value to allow.
+            alpha (`float`, optional): Real number between 0 and 1 used to
```
The accepted types don't match the documentation in the method definition.

joaander commented on hoomd/tune/solve.py:
```diff
@@ -0,0 +1,516 @@
+        if tunable not in self._bins:
+            self._initial_binning(tunable)
+            tunable.x = self._get_bin_center(tunable, 0)
+            return
```
Shouldn't this return `True` or `False`?

joaander commented on hoomd/tune/solve.py:
```diff
@@ -0,0 +1,516 @@
+                bin_y.clear()
+                tunable.x = self._get_bin_center(tunable, 0)
+                self._round[tunable] += 1
+            return
```
Shouldn't this return `True` or `False`?

joaander commented on hoomd/tune/solve.py:
```diff
@@ -0,0 +1,516 @@
+        return self._a
+
+
+class BinaryOptimizer(Optimizer):
```
Binary searches work only on monotonic input which we are not likely to ever provide this. Also, it is a binary search only when `n_bins=1` and `n_rounds > 1`.

Rename this to `BucketOptimizer` or `GridOptimizer` to better indicate the behavior of this code. Also, set an appropriate default number of bins - e.g. 5.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,281 @@
+        )
+
+    @classmethod
+    def with_binary(
```
Rename to match the new name of the optimizer. Match `n_bins` and `n_rounds` defaults between the two codes.

b-butler commented:
@joaander I think it is ready for another round of reviews now.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,281 @@
+        return self._tuned
+
+    @hoomd.logging.log
+    def max_tps(self):
```
`max_tps` doesn't update after the first time it is set (see example output in review comment).

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,281 @@
+        return self._max_tps
+
+    @hoomd.logging.log
+    def best_buffer_size(self):
```
`best_buffer_size` doesn't update after the first time it is set.

joaander reviewed (changes requested):
Thanks, the new optimizer works well and consistently identifies a good buffer value.

Here is the output from my latest test (attached):
```
▶ python run.py
Initial buffer: 0.45
TPS: 2911.7286175753106

Start grid search tuning
     buffer          last_tps         max_tps      best_buffer_size
    0.10000          0.00000          0.00000          0.10000     
    0.10000         1891.65436       1891.65436        0.10000     
    0.30000         1883.85896       1891.65436        0.10000     
    0.50000         2767.69185       1891.65436        0.10000     
    0.70000         2900.46587       1891.65436        0.10000     
    0.90000         2748.63159       1891.65436        0.10000     
    0.42000         2571.41395       1891.65436        0.10000     
    0.46000         2749.95057       1891.65436        0.10000     
    0.50000         2799.36007       1891.65436        0.10000     
    0.54000         2855.26980       1891.65436        0.10000     
    0.58000         2858.78462       1891.65436        0.10000     
    0.54000         2704.27996       1891.65436        0.10000     
    0.54000         2853.00111       1891.65436        0.10000     
Tuned: True
     buffer     
    0.54000     
TPS: 2846.3956803099154

Start gradient descent tuning
     buffer          last_tps         max_tps      best_buffer_size
    0.10000          0.00000          0.00000          0.10000     
    0.11000         1810.07368       1810.07368        0.10000     
    0.10276         1795.73319       1810.07368        0.10000     
    0.09215         1804.15273       1810.07368        0.10000     
    0.08446         1807.29010       1810.07368        0.10000     
    0.07898         1811.22310       1810.07368        0.10000     
    0.07303         1815.32021       1810.07368        0.10000     
    0.07385         1810.44955       1810.07368        0.10000     
    0.07608         1810.57411       1810.07368        0.10000     
    0.08559         1813.95859       1810.07368        0.10000     
    0.08636         1805.52965       1810.07368        0.10000     
    0.07827         1804.35488       1810.07368        0.10000     
    0.07180         1804.36042       1810.07368        0.10000     
    0.07930         1791.52523       1810.07368        0.10000     
    0.09377         1803.30618       1810.07368        0.10000     
    0.09905         1794.22951       1810.07368        0.10000     
    0.09789         1792.46805       1810.07368        0.10000     
    0.10696         1789.88504       1810.07368        0.10000     
    0.11071         1784.45975       1810.07368        0.10000     
    0.11097         1783.40096       1810.07368        0.10000     
Tuned: True
    0.11097         1783.40096       1810.07368        0.10000     
TPS: 1784.408799490801
```
[run.txt](https://github.com/glotzerlab/hoomd-blue/files/8702717/run.txt)

b-butler committed:
refactor: Move TPS/walltime broadcast to getter

This should speed up the performance of System::run by avoiding two
broadcast calls.

b-butler committed:
doc: Improve solve subsystem documentation

b-butler committed:
doc: Improve documentation of Tuner.tuned properties

b-butler committed:
feat: Add __eq__ for GridOptimizer

b-butler committed:
test: Test NeighborListBuffer.with_grid.

b-butler committed:
fix: Fix NeighborListBuffer.tuned behavior to match documentation

b-butler committed:
refactor: Do not scale TPS in NeighborListBuffer tuner

Leave it to the different solvers to handle gradient size.

b-butler committed:
doc: Fix documenation of GradientDescent.kappa.

b-butler committed:
doc: Document GradientDescent attributes

b-butler committed:
feat: Make GradientDescent.alpha a variant.

This lets alpha vary across the solving.

b-butler committed:
refactor: NeighborListBuffer.with_gradient_descent alpha default

New default uses the new ability to use a variant for alpha.

b-butler committed:
fix: remove errant const

b-butler committed:
feat: Add final_bin property for NeighborListBuffer tuner

When using a GridOptimizer solver, NeighborListBuffer now has a property
final_bin that after tuning returns the bin boundaries which had the
optimal TPS.

b-butler committed:
feat: Allow setting of minimum_buffer in NeighborListBuffer.

b-butler committed:
doc: Fix lingering reference to NeighborListBuffer.with_binary

b-butler committed:
Merge branch 'trunk-minor' into feature/nlist-buffer-tuner

b-butler committed:
Merge branch 'trunk-minor' into feature/nlist-buffer-tuner

b-butler commented:
@joaander and @Charlottez112 the PR is ready for another round of reviews. I have attempted to make `GradientDescent` more useful by allowing `alpha` to be a variant, but it still inconsistently finds the optimum.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,322 @@
+    Warning:
+        When using with a `hoomd.device.GPU` device, autotuning can mess with
+        convergence. Running the simulation before adding the tuner to allow the
+        autotuning to converge first results in better TPS optimization.
```
```suggestion
        When using with a `hoomd.device.GPU` device, kernel launch parameter
        autotuning can prevent convergence. Run the simulation for 25,000 steps
        before adding the tuner to allow the autotuning to converge first
        results in better TPS optimization.
```
Add some specificity.

joaander commented on hoomd/md/tune/nlist_buffer.py:
```diff
@@ -0,0 +1,322 @@
+    Direct instantiation of this class requires a `hoomd.tune.Optimizer`
+    that determines how move sizes are updated. This class also provides class
+    methods to create a `NeighborListBuffer` tuner with built-in solvers; see
+    `NeighborListBuffer.with_grid` and
```
Could we put this in a `Tip:` block to make it more obvious? I've gotten feedback from some users that tried (and failed) to use the MC move size tuners with the documented constructor and missed the helpful `with_` classmethods.

joaander commented on hoomd/simulation.py:
```diff
@@ -345,6 +345,18 @@ def final_timestep(self):
 
+    @log
+    def initial_timestep(self):
+        """float: `run` will end at this timestep.
```
```suggestion
        """float: `run` started at this timestep.
```

joaander reviewed (changes requested):
Thanks!

I've tested this and I can get consistent results using `with_grid` with various values for `n_bins` and `n_rounds`). At least as consistent as you can get when performance regularly fluctuates by ~1% and the average performance only varies ~1% over a large range of buffers. I tested both GPU and MPI with CPUs.

I trust your testing of the gradient descent tuner.

I have a few suggestions to improve the documentation, then this is good to merge. A `reset()` method to start the grid search over again would be nice, but not required.

b-butler committed:
doc: Add specificity to nlist buffer tuning on GPUs

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
doc: Fix docstring copying error

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
feat: Add reset method to solvers to reset solving.

b-butler committed:
feat: Add reset method to NeighborListBuffer.

Always tuning to begin again.

b-butler committed:
doc: Use tip to highlight NeighborListBuffer factory methods

joaander reviewed (approved):
Thanks!

Are any final tweaks needed before merging?

b-butler committed:
doc: Use tip block for HPMC tuner factory methods

b-butler commented:
@joaander now I am done. Just added the `Tip:` block to the other factory methods for tuners.
